packages/ember-glimmer/tests/integration/components/dynamic-components-test.js

```diff
@@ -292,4 +308,334 @@ moduleFor('Components test: dynamic components', class extends RenderingTest {
     assert.deepEqual(destroyed, { 1: 1, 2: 1, 3: 1, 4: 1, 5: 1, 6: 1, 7: 1, 8: 1 });
   }
 
+  ['@test component helper destroys underlying component when it is swapped out'](assert) {
+    let destroyed = { 'foo-bar': 0, 'foo-bar-baz': 0 };
+
+    this.registerComponent('foo-bar', {
+      template: 'hello from foo-bar',
+      ComponentClass: Component.extend({
+        willDestroy() {
+          this._super();
+          destroyed['foo-bar']++;
+        }
+      })
+    });
+
+    this.registerComponent('foo-bar-baz', {
+      template: 'hello from foo-bar-baz',
+      ComponentClass: Component.extend({
+        willDestroy() {
+          this._super();
+          destroyed['foo-bar-baz']++;
+        }
+      })
+    });
+
+    this.render('{{component componentName name=name}}', { componentName: 'foo-bar' });
+
+    assert.deepEqual(destroyed, { 'foo-bar': 0, 'foo-bar-baz': 0 });
+
+    this.runTask(() => this.rerender());
+
+    assert.deepEqual(destroyed, { 'foo-bar': 0, 'foo-bar-baz': 0 });
+
+    this.runTask(() => set(this.context, 'componentName', 'foo-bar-baz'));
+
+    assert.deepEqual(destroyed, { 'foo-bar': 1, 'foo-bar-baz': 0 });
+
+    this.runTask(() => set(this.context, 'componentName', 'foo-bar'));
+
+    assert.deepEqual(destroyed, { 'foo-bar': 1, 'foo-bar-baz': 1 });
+  }
+
+  ['@test component helper with bound properties are updating correctly in init of component'](assert) {
+    this.registerComponent('foo-bar', {
+      template: 'foo-bar {{location}} {{yield}}',
+      ComponentClass: Component.extend({
+        init: function() {
+          this._super(...arguments);
+
+          assert.equal(this.get('location'), 'Caracas', 'location is bound on init');
```

Comment:
I think we should either...

```
let initedLocation;

...

..., {
  init: function() {
    this._super(...arguments);
    initedLocation = get(this, 'location'));
```

and assert it in the regular flow of the test, or

```
  init: function() {
    this._super(...arguments);
    this.locationCopy = get(this, 'location');
``` 

...and use `{{locationCopy}}` in the template (and assert it with `assertText`).

In the current form, you won't really notice if this assertion is not executed.